storage/cache/subtree_cache.go

```diff
@@ -51,20 +55,24 @@ const (
 )
 
 // NewSubtreeCache returns a newly intialised cache ready for use.
+// populateSubtree is a function which knows how to populate a subtree's
+// internal nodes given its leaves, and will be called for each subtree loaded
+// from storage.
 // TODO(al): consider supporting different sized subtrees - for now everything's subtrees of 8 levels.
-func NewSubtreeCache() SubtreeCache {
+func NewSubtreeCache(populateSubtree storage.PopulateSubtreeFunc) SubtreeCache {
 	return SubtreeCache{
-		subtrees:      make(map[string]*storage.SubtreeProto),
-		dirtyPrefixes: make(map[string]bool),
-		mutex:         new(sync.RWMutex),
+		subtrees:        make(map[string]*storage.SubtreeProto),
+		dirtyPrefixes:   make(map[string]bool),
+		mutex:           new(sync.RWMutex),
+		populateSubtree: populateSubtree,
 	}
 }
 
 // splitNodeID breaks a NodeID out into its prefix and suffix parts.
 // unless ID is 0 bits long, Suffix must always contain at least one bit.
 func splitNodeID(id storage.NodeID) ([]byte, Suffix) {
 	if id.PrefixLenBits == 0 {
-		return []byte{}, Suffix{bits: 0, path: []byte{}}
+		return []byte{}, Suffix{bits: 0, path: []byte{0}}
```

Comment:
Slightly surprised that an empty path isn't empty but you obviously changed this for a reason.